blocks/subscriptions-block/components/usePaywall.js

```diff
@@ -0,0 +1,122 @@
+  const [campaignCode, setCampaignCode] = useState();
+  const [results, setResults] = useState();
+  const [triggeredRule, setTriggeredRule] = useState();
+  const [otherTriggeredRules, setOtherTriggeredRules] = useState([]);
+  const [isSignwall, setIsSignwall] = useState();
+
+  // eslint-disable-next-line no-underscore-dangle
+  const rules = (!isServerSide() && window?.ArcP?._rules) || [];
+
+  const apiOrigin = api?.retail?.origin;
+
+  useEffect(() => {
+    const runPaywall = async () => {
+      setResults(
+        // Subs ArcP.run assumes https://, so we need to strip it from the endpoint origin.
+        await window?.ArcP?.run({
+          apiOrigin: apiOrigin.replace(/^https?:\/\//i, ''),
+          contentIdentifier: globalContent.canonical_url,
+          contentRestriction: globalContent.content_restrictions.content_code,
+          contentType: globalContent.type,
+          Identity,
+          paywallFunction: (campaign) => {
+            setCampaignCode(campaign);
+            setIsPaywalled(true);
+          },
+          section: globalContent.taxonomy.primary_section._id,
+        }),
+      );
+    };
+    if (
+      apiOrigin
+      && globalContent.canonical_url
+      && Identity
+      && !isPaywalled
+      && !isServerSide()
+    ) {
+      runPaywall();
+    }
+  }, [
+    apiOrigin,
+    globalContent.canonical_url,
+    globalContent.content_restrictions.content_code,
+    globalContent.taxonomy.primary_section._id,
+    globalContent.type,
+    isPaywalled,
+  ]);
+
+  useEffect(() => {
+    if (results && rules?.length) {
+      const newTriggeredRule = rules?.find((rule) => results?.triggered?.id === rule.id);
+      const newOtherTriggeredRules = rules?.filter((rule) => rule.rt[1] < results?.triggered?.rc);
+      if (newTriggeredRule) {
+        setTriggeredRule(newTriggeredRule);
+      }
+      if (newOtherTriggeredRules?.length) {
+        setOtherTriggeredRules(newOtherTriggeredRules);
+      }
+    }
+  }, [results, rules]);
+
+  useEffect(() => {
+    if (triggeredRule || otherTriggeredRules) {
+      if (triggeredRule?.e?.length > 1) {
+        setIsSignwall(false);
+        return;
+      }
+      if (otherTriggeredRules?.length) {
+        const paywallRule = otherTriggeredRules
+          .sort((a, b) => b.rt[1] - a.rt[1])
+          .find((rule) => rule?.e?.length > 1);
+        setIsSignwall(!paywallRule);
+        if (paywallRule && triggeredRule !== paywallRule) {
+          setTriggeredRule(paywallRule);
+        }
+        return;
+      }
+      if (triggeredRule?.e?.length === 1) {
+        setIsSignwall(true);
+      }
+    }
+  }, [triggeredRule, otherTriggeredRules]);
+
+  const getRegistrationRules = useCallback(
+    () => rules?.filter((rule) => rule.e.length === 1 && rule.e[0]),
+    [rules],
+  );
+
+  const getSubscribeRules = useCallback(
+    () => rules?.filter((rule) => rule.e.length > 1),
+    [rules],
+  );
+
+  if (isServerSide()) {
+    return null;
```

Comment:
This is going to cause issues in parent functions as if you are using the hook in the parent like 

```
const { results } = usePaywall();
```

`results` is going to be undefined and throw an error - can we always return the same named object keys?